packages/stl-serializer/CSGToStla.js

```diff
@@ -1,41 +1,62 @@
+const { ensureManifoldness } = require('@jscad/io-utils')
+const {isCSG} = require('@jscad/csg')
 
-function serialize (CSG, options) {
-  options && options.statusCallback && options.statusCallback({progress: 0})
-  var result = 'solid csg.js\n'
-  CSG.polygons.map(function (p, i) {
-    result += CSGPolygontoStlString(p)
-    options && options.statusCallback && options.statusCallback({progress: 100 * i / CSG.polygons.length})
-  })
-  result += 'endsolid csg.js\n'
-  options && options.statusCallback && options.statusCallback({progress: 100})
+// objects must be an array of CSG objects (with polygons)
+const serializeText = (objects, options) => {
+  options.statusCallback && options.statusCallback({progress: 0})
+
+  let result = `solid JSCAD
+${convertToStl(objects, options)}
+endsolid JSCAD
+`
+  options.statusCallback && options.statusCallback({progress: 100})
   return [result]
 }
 
-function CSGVector3DtoStlString (v) {
-  return v._x + ' ' + v._y + ' ' + v._z
+const convertToStl = (objects, options) => {
+  let result = []
+  objects.forEach(function (object, i) {
+    result.push(convertToFacets(object, options))
+    options.statusCallback && options.statusCallback({progress: 100 * i / objects.length})
+  })
+  return result.join('\n')
+}
+
+const convertToFacets = (object, options) => {
+  let result = []
+  object.polygons.forEach(function (p, i) {
+    result.push(convertToFacet(p))
+  })
+  return result.join('\n')
+}
+
+const vector3DtoStlString = (v) => {
+  return `${v._x} ${v._y} ${v._z}`
 }
 
-function CSGVertextoStlString (vertex) {
-  return 'vertex ' + CSGVector3DtoStlString(vertex.pos) + '\n'
+const vertextoStlString = (vertex) => {
+  return `vertex ${vector3DtoStlString(vertex.pos)}`
 }
 
-function CSGPolygontoStlString (polygon) {
-  var result = ''
+const convertToFacet = (polygon) => {
+  let result = []
   if (polygon.vertices.length >= 3) {
-    // STL requires triangular polygons. If our polygon has more vertices, create
-    // multiple triangles:
-    var firstVertexStl = CSGVertextoStlString(polygon.vertices[0])
-    for (var i = 0; i < polygon.vertices.length - 2; i++) {
-      result += 'facet normal ' + CSGVector3DtoStlString(polygon.plane.normal) + '\nouter loop\n'
-      result += firstVertexStl
-      result += CSGVertextoStlString(polygon.vertices[i + 1])
-      result += CSGVertextoStlString(polygon.vertices[i + 2])
-      result += 'endloop\nendfacet\n'
+    // STL requires triangular polygons. If our polygon has more vertices, create multiple triangles:
```

Comment:
I think we actually have a triangulate() or similar function somewhere in CSG.js , will try to find it (not important for this PR / version though)

Reply:
Correct. CSG has toTriangles() but really polygon3D should have this function. Also, there are several methods to split a polygon into triangles as well.